src/psyclone/parse/file_info.py

```diff
@@ -77,27 +143,286 @@ def filename(self):
+        except FileNotFoundError:
+            if verbose:
+                print(f"  - No cache file '{self._filepath_cache}' found")
+            return None
+
+        # Unpack cache file
+        try:
+            cache: _CacheFileInfo = pickle.load(filehandler)
+        except Exception as ex:
+            print("Error while reading cache file - ignoring: " + str(ex))
+            return None
+
+        # Verify checksums
+        if cache._source_code_hash_sum != self._source_code_hash_sum:
+            if verbose:
+                print(
+                    f"  - Cache hashsum mismatch: "
+                    f"source {self._source_code_hash_sum} "
+                    f"vs. cache {cache._source_code_hash_sum}"
+                )
+            return None
+
+        for key in ["_source_code_hash_sum", "_fparser_tree", "_psyir_node"]:
+            assert key in cache.__dict__.keys()
+
+        self._cache_data_load = cache
+        self._source_code_hash_sum = \
+            self._cache_data_load._source_code_hash_sum
+
+    def _cache_save(
+        self,
+        verbose: bool = False,
+    ) -> None:
+        """Save the following elements to a cache file:
+        - hash sum of code
+        - fparser tree
+        - potentially psyir nodes
+
+        :param verbose: Produce some verbose output
+        """
+
+        if not self._use_caching:
+            return None
+
+        if self._source_code_hash_sum is None:
+            # Nothing to cache
+            return None
+
+        cache_updated = False
+        if self._cache_data_save is None:
+            # Cache doesn't exist => prepare data to write to file
+            self._cache_data_save = _CacheFileInfo()
+            self._cache_data_save._source_code_hash_sum = (
+                self._source_code_hash_sum)
+
+        assert self._cache_data_save._source_code_hash_sum == (
+                    self._source_code_hash_sum)
+
+        if (
+            self._cache_data_save._fparser_tree is None
+            and self._fparser_tree is not None
+        ):
+            # Make copies of it since they could be modified later
+            # With this, we can also figure out potential issues with
+            # the serialization in fparser
+            self._cache_data_save._fparser_tree = \
+                copy.deepcopy(self._fparser_tree)
+            cache_updated = True
+
+        if self._cache_data_save._psyir_node is None and (
+                self._psyir_node is not None):
+            # TODO #2786: Serialization of psyir tree not possible
+            #
+            # E.g., this call fails: copy.deepcopy(self._psyir_node)
+            #
+            # Uncomment this code if serialization of psyir tree is
+            # possible and it will work.
+            # self._cache._psyir_node = copy.deepcopy(self._psyir_node)
+            # cache_updated = True
+            pass
+
+        if not cache_updated:
+            return None
+
+        # Open cache file
```

Comment:
This bit will need to be atomic to support parallel builds. When creating output files in another place we do:
https://github.com/stfc/PSyclone/blob/28f7608f19d99f938584d8bf474314ce26155798/src/psyclone/psyGen.py#L1647-L1667
For now though, as long as we default to not using caching then this can be left as is, albeit with a comment to this effect.

Reply:
I see from googling that this requirement for POSIX compliance could be problematic on both NFSv4 and Lustre but is there in theory:
![image](https://github.com/user-attachments/assets/ed55560c-8961-45c2-9bd9-3a15bfee7cfc)
(from https://wiki.lustre.org/NFS_vs._Lustre).

Reply:
Wonderful! It's what I was searching for. Thanks. It's now done like that.

Reply:
I now understood your trick with either creating the file and owning it or raising an Exception.
I added this version. But I also have to remove the file if it exists.
There can still be race conditions, but in the worst case, the cache file will be accidentally deleted.